### DISCUSSION TITLE
Issues about Logger in PaddleNLP

### DISCUSSION THREAD

**Discussion by wj-Mcat** [PaddlePaddle/PaddleNLP#3689] [General]
## 背景

PaddleNLP中的logger在最初设计时是支持: `logger.train`, `logger.eval` 以及logger常规（`info`, `error`）的用法，可是在本项目当中基本上是没用过的，其中一个原因在于：logger没有`代码提示`。（在IDE中输入`.`就可以出现logger对象的所有属性）

## 问题

当前的logger对象本质上是通过给自定义`Logger`对象绑定不同方法实现的，核心代码如下所示：

https://github.com/PaddlePaddle/PaddleNLP/blob/e785bb116341d35f144bbf9d5dcbbf92c96cb233/paddlenlp/utils/log.py#L71-L93

存在以下问题：
1. 通过动态绑定，导致logger对象无法实现代码智能提示。
2. 框架当中没有使用`.train/.eval`等方法，后续的编码过程中可普及。
3. 框架当中普遍使用info,warning，对于debug方法的使用很少，除了`pipelines`中的模块，同时loglevel的使用也是比较随意。可参考： [syslog-Severity_level](https://en.wikipedia.org/wiki/Syslog#Severity_level), [Understanding Logging Levels: What They Are & How To Use Them](https://sematext.com/blog/logging-levels/)
4. 无法给框架logger添加FileHandler（代码写死，没提供扩展的空间）
5. Logger 打印出的信息可更详细一点，比如添加执行的模块名称、函数名称等

## 可优化的点

1. 在保证logger正常接口的同时，将`Logger`类调整成为logger.Logger的子类，同时支持自定义train/eval等方法
2. 添加自定义FileHandler的扩展接口，保证可将框架的训练日志都保存到日志文件当中。
3. 细化logger format，添加细粒度的日志信息
4. 将logger默认level调整成为`INFO`, 同时还支持自定义设置（通过环境变量设置） logger level。

**Comment by ZHUI** [comment]
之前 hf 的 trainer 也有部分日志的内容，我没有迁移过来。可以考虑一并迁移。
另外的话，分布式启动的话，日志会重定向`log/worklog.0` 之类的，注意看看会不会影响我们。

**Reply by wj-Mcat** [reply]
关于重定向的问题，能否提供简单的复现代码，我这边也可以做有针对性的测试和调整。 @ZHUI

**Reply by ZHUI** [reply]
分布式训练，默认就有。
